### PR TITLE
chore: add fuzzy tests

### PR DESCRIPTION
**Pull request by jonybur**
# Description

Fuzzy tests for `noir_js`.

Checks that the different methods of noir_js fail gracefully given a random input.

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/2888

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/integration-tests/test/fuzzy/fuzzy.test.ts
Comment (TomAFrench) [inline compiler/integration-tests/test/fuzzy/fuzzy.test.ts:1]: None of these are integration tests and should be moved to cover the relevant packages.

File: compiler/integration-tests/test/fuzzy/fuzzy.test.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+describe("compile fuzz test", () => {
+  for (let i = 0; i < iterations; i++) {
+    it("should error compile() gracefully", async () => {
+      await newCompiler();
+      await newABICoder();
+      await initACVM();
+      const fuzzedNoirSource = getRandomString(
+        Math.floor(Math.random() * Math.pow(10, i)),
+      );
```
Comment (TomAFrench) [inline compiler/integration-tests/test/fuzzy/fuzzy.test.ts:21]: I don't think that this is a particularly useful test. This input is going to look nothing like Noir source code so what's the benefit of running this test multiple times?

It would be better to run some of out `compile_failure` tests on the wasm as they're closer to meaningful failure cases.

File: compiler/integration-tests/test/fuzzy/fuzzy.test.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+const { default: newABICoder, abiEncode } = abi;
+
+const iterations = 5;
+
+describe("compile fuzz test", () => {
+  for (let i = 0; i < iterations; i++) {
+    it("should error compile() gracefully", async () => {
+      await newCompiler();
+      await newABICoder();
+      await initACVM();
```
Comment (TomAFrench) [inline compiler/integration-tests/test/fuzzy/fuzzy.test.ts:18]: These are unused in this test

File: compiler/integration-tests/test/fuzzy/fuzzy.test.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+// timeouts and errors in acvm_js
+xdescribe("executeCircuit fuzz test", function () {
+  this.timeout(120e3);
+
+  for (let i = 0; i < iterations; i++) {
+    it(`should error executeCircuit() gracefully`, async () => {
+      let errorTriggered = false;
+      const numberLength = Math.pow(10, i);
+      try {
+        const compressedByteCode = getRandomUint8Array(numberLength);
```
Comment (TomAFrench) [inline compiler/integration-tests/test/fuzzy/fuzzy.test.ts:72]: Again, we'll always fail with the same error due to this not deserializing into valid ACIR. Why not just add a test where we pass in some invalid bytes once rather than multiple times?

File: compiler/integration-tests/test/fuzzy/fuzzy.test.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  for (let i = 0; i < iterations; i++) {
+    it(`should error abiEncode() gracefully`, async () => {
+      let errorTriggered = false;
+      const stringLength = Math.pow(10, i);
+      const randomString = getRandomString(stringLength);
+      const fuzzedTOML = `${randomString} = "${randomString}"`;
+      try {
+        const inputs = TOML.parse(fuzzedTOML);
+        expect(inputs).to.be.an("object");
+        abiEncode(randomString, inputs, null);
```
Comment (TomAFrench) [inline compiler/integration-tests/test/fuzzy/fuzzy.test.ts:51]: This will always fail as you're passing in a string rather than an actual ABI object as the ABI. We could then replace this with a single test that we throw an error when we encounter an invalid ABI.

File: compiler/integration-tests/test/fuzzy/fuzzy.test.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      } catch (e) {
+        expect(e.toString().includes("RuntimeError: unreachable executed")).to
+          .be.true;
+        errorTriggered = true;
+      }
+
+      expect(errorTriggered).to.be.true;
+    });
+  }
+});
```
Comment (TomAFrench) [inline compiler/integration-tests/test/fuzzy/fuzzy.test.ts:106]: This is equivalent to fuzzing the trait implementation below
https://github.com/noir-lang/noir/blob/4deb07f80ce110187b66a46dd5624af3b8df3dbd/acvm-repo/acir/src/native_types/witness_map.rs#L81-L91

If we wanted to fuzz this then I think it would be more effective to do this inside Rust tbh.

**Review by TomAFrench — CHANGES_REQUESTED** [review]
In general, when fuzzing we want to generate inputs which closely match the expected supported input space so failures are unexpected and correspond to unhandled edge cases. Here however we're generating invalid inputs 99.99% of the time so we don't gain any insights.

Due to the inherent structure required in the inputs, these tests are basically all going to fail due to deserialisation/lexing errors. We could then make these tests simpler and more useful by targetting them at the failure cases we want to trigger.

File: compiler/integration-tests/test/fuzzy/fuzzy.test.ts (hunk 5 above)
Comment (TomAFrench) [reply compiler/integration-tests/test/fuzzy/fuzzy.test.ts:106]: That said, I don't think this would be a great use of time as we're essentially just fuzz testing `GzEncoder` whereas we can just rely on their test suite.

**Comment by kevaundray** [comment]
I think if this is meant for fuzzing noir_js then the fuzzing code should be in the noir_js package
